include/kwk/context/sycl/context.hpp

```diff
@@ -0,0 +1,441 @@
+
+          bool* filter = ::sycl::malloc_device<bool>(numel, *(this));
+
+          parent::submit([&](::sycl::handler& h)
+          {
+            ::sycl::accessor data_acc = in_proxy.access(h);
+
+            // First, test all values
+            h.parallel_for( ::sycl::range<1>(numel)
+                          , [=](::sycl::id<1> i)
+                            {
+                              if (func(data_acc[i[0]])) filter[i[0]] = true;
+                              else                      filter[i[0]] = false;
+                            }
+                          );
+          });
+          parent::wait();
+
+          if (ascending)
+          {
+            parent::submit([&](::sycl::handler& h)
+            {
+              auto reduction_obj = ::sycl::reduction(result_buf, h, rfunc_ascending);
+
+              // Then reduce the elements
+              h.parallel_for( ::sycl::range<1>(numel)
+                            , reduction_obj
+                            , [=](::sycl::id<1> i, auto& sum)
+                              {
+                                sum.combine(filter[i[0]] ? i[0] : -1);
+                              }
+                            );
+            });
+          }
+          else
+          {
+            parent::submit([&](::sycl::handler& h)
+            {
+              auto reduction_obj = ::sycl::reduction(result_buf, h, rfunc_descending);
+
+              // Then reduce the elements
+              h.parallel_for( ::sycl::range<1>(numel)
+                            , reduction_obj
+                            , [=](::sycl::id<1> i, auto& sum)
+                              {
+                                sum.combine(filter[i[0]] ? i[0] : -1);
+                              }
+                            );
+            });
+          }
+
+          parent::wait();
+          ::sycl::free(filter, *this);
+        }
+      }
+      catch (::sycl::exception const &e)
+      {
+        std::cout << "An exception is caught for SYCL transform_reduce.\n";
+        std::terminate();
+      }
+
+      if (result != -1)
+      {
+        // TODO: change to kwk::coordinates(it, out.shape());
+        return std::optional<kwk::position<In::static_order>>{kwk::utils::tools::linear_to_pos(result, in)};
+      } else {
+        return std::nullopt;
+      }
+    }
+
+
+    std::string get_device_name()
+    {
+      return parent::get_device().get_info<::sycl::info::device::name>();
+    }
+
+
+    bool is_gpu()
+    {
+      return parent::get_device().is_gpu();
+    }
+  };
+
+  // Checks whether GPU offloading is available in this system
+  // Don't forget the compilation flag "-fsycl -fsycl-targets=nvptx64-nvidia-cuda,x86_64" x86_64 or spir64
+  // to target both Nvidia GPUs and native CPUs.
+  [[maybe_unused]] static bool has_gpu()
+  {
+    auto devices = ::sycl::device::get_devices();
+    for (const auto& dev : devices)
```

Comment:
Looks interesrting, would you mind dropping an issue on SPY to add it there ?

Reply:
Sure! Here is the issue: https://github.com/jfalcou/spy/issues/64